Mathlib/RingTheory/Spectrum/Prime/Topology.lean

```diff
@@ -862,26 +883,73 @@ lemma existsUnique_idempotent_basicOpen_eq_of_isClopen {s : Set (PrimeSpectrum R
-  · rw [PrimeSpectrum.basicOpen_eq_zeroLocus_compl, Set.compl_subset_comm, ← hI']
-    exact PrimeSpectrum.zeroLocus_anti_mono
-      (Set.singleton_subset_iff.mpr <| Ideal.pow_le_self hnz hx)
+  obtain ⟨x, hx, y, hy, add⟩ := this
+  have mul : x * y = 0 := hn (Ideal.mul_mem_mul hx hy)
+  have : s = basicOpen x := by
+    refine subset_antisymm ?_ ?_
+    · rw [← hJ', basicOpen_eq_zeroLocus_of_mul_add _ _ mul add]
+      exact zeroLocus_anti_mono (Set.singleton_subset_iff.mpr <| Ideal.pow_le_self hnz hy)
+    · rw [basicOpen_eq_zeroLocus_compl, Set.compl_subset_comm, ← hI']
+      exact zeroLocus_anti_mono (Set.singleton_subset_iff.mpr <| Ideal.pow_le_self hnz hx)
+  refine ⟨x, y, mul, add, this, ?_⟩
+  rw [this, basicOpen_eq_zeroLocus_of_mul_add _ _ mul add, basicOpen_eq_zeroLocus_compl]
 
 lemma exists_idempotent_basicOpen_eq_of_isClopen {s : Set (PrimeSpectrum R)}
     (hs : IsClopen s) : ∃ e : R, IsIdempotentElem e ∧ s = basicOpen e :=
-  (existsUnique_idempotent_basicOpen_eq_of_isClopen hs).exists
+  have ⟨e, _, mul, add, eq, _⟩ := exists_mul_eq_zero_add_eq_one_basicOpen_eq_of_isClopen hs
+  ⟨e, (IsIdempotentElem.of_mul_add mul add).1, eq⟩
+
+@[stacks 00EE]
+lemma existsUnique_idempotent_basicOpen_eq_of_isClopen {s : Set (PrimeSpectrum R)}
+    (hs : IsClopen s) : ∃! e : R, IsIdempotentElem e ∧ s = basicOpen e := by
+  refine existsUnique_of_exists_of_unique (exists_idempotent_basicOpen_eq_of_isClopen hs) ?_
+  rintro x y ⟨hx, rfl⟩ ⟨hy, eq⟩
+  exact basicOpen_injOn_isIdempotentElem hx hy (SetLike.ext' eq)
 
 @[deprecated (since := "2024-11-11")]
 alias exists_idempotent_basicOpen_eq_of_is_clopen := exists_idempotent_basicOpen_eq_of_isClopen
 
+open TopologicalSpace.Opens in
+lemma isClopen_iff_mul_add {s : Set (PrimeSpectrum R)} :
+    IsClopen s ↔ ∃ e f : R, e * f = 0 ∧ e + f = 1 ∧ s = basicOpen e := by
+  refine ⟨fun h ↦ ?_, ?_⟩
+  · have ⟨e, f, h⟩ := exists_mul_eq_zero_add_eq_one_basicOpen_eq_of_isClopen h
+    exact ⟨e, f, by simp only [h, and_self]⟩
+  rintro ⟨e, f, mul, add, rfl⟩
+  exact isClopen_basicOpen_of_mul_add e f mul add
+
+lemma isClopen_iff_mul_add_zeroLocus {s : Set (PrimeSpectrum R)} :
+    IsClopen s ↔ ∃ e f : R, e * f = 0 ∧ e + f = 1 ∧ s = zeroLocus {e} := by
+  rw [isClopen_iff_mul_add, exists_swap]
+  refine exists₂_congr fun e f ↦ ?_
+  rw [mul_comm, add_comm, ← and_assoc, ← and_assoc, and_congr_right]
+  intro ⟨mul, add⟩
+  rw [zeroLocus_eq_basicOpen_of_mul_add e f mul add]
+
+open TopologicalSpace (Clopens)
+
+/-- Clopen subsets in the prime spectrum of a commutative semiring are in order-preserving
+bijection with pairs of elements with product 0 and sum 1. (By definition, `(e₁, f₁) ≤ (e₂, f₂)`
+iff `e₁ * e₂ = e₁`.) Both elements in such pairs must be idempotents, but there may exists
+idempotents that do not form such pairs (does not have a "complement"). For example, in the
+semiring {0, 0.5, 1} with ⊔ as + and ⊓ as *, 0.5 has no complement. -/
+def mulZeroAddOneEquivClopens :
```

Comment:
Can you add this (and maybe something else) to the docstring at the beginning of the file?

Reply:
LGTM, just please add the doc at the beginning of the file. Thanks!

Reply:
I added ~60 lines of docstring. Would you like to take a look?

Reply:
Thanks a lot for doing that, that's perfect!